src/mlabns/mlabns_impl.hpp

```diff
@@ -0,0 +1,114 @@
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "src/ext/json/src/json.hpp"
+#include <measurement_kit/common.hpp>
+#include <measurement_kit/http.hpp>
+#include <measurement_kit/mlabns.hpp>
+#include <regex>
+#include <string>
+#include <utility>
+
+using json = nlohmann::json;
+
+namespace mk {
+namespace mlabns {
+
+static inline ErrorOr<std::string> as_query(Settings &settings) {
+    std::string query;
+    std::string policy = settings.get<std::string>("mlabns/policy", "");
+    std::string metro = settings.get<std::string>("mlabns/metro", "");
+    std::string address_family =
+        settings.get<std::string>("mlabns/address_family", "");
+    if (policy == "" && metro == "" && address_family == "") {
+        return query;
+    }
+    if (policy != "") {
+        if (policy != "geo" && policy != "random" && policy != "metro" &&
+            policy != "country") {
+            return InvalidPolicyError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "policy=" + policy;
+    }
+    if (metro != "") {
+        std::regex valid_metro("^[a-z]{3}$");
+        if (!std::regex_match(metro, valid_metro)) {
+            return InvalidMetroError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "metro=" + metro;
+    }
+    if (address_family != "") {
+        if (address_family != "ipv4" && address_family != "ipv6") {
+            return InvalidAddressFamilyError();
+        }
+        if (query != "") {
+            query += "&";
+        }
+        query += "address_family=" + address_family;
+    }
+    query = "?" + query;
+    return query;
+}
+
+template <MK_MOCK_NAMESPACE(http, get)>
+inline void query_impl(std::string tool, Callback<Error, Reply> callback,
+                        Settings settings, Var<Reactor> reactor,
+                        Var<Logger> logger) {
+    ErrorOr<std::string> query = as_query(settings);
+    if (!query) {
+        callback(query.as_error(), Reply());
+        return;
+    }
+    std::string url = "https://mlab-ns.appspot.com/";
+    std::regex valid_tool("^[a-z]+$");
+    if (!std::regex_match(tool, valid_tool)) {
+        callback(InvalidToolNameError(), Reply());
+        return;
+    }
+    url += tool;
+    url += *query;
+    logger->info("query mlabns for tool %s", tool.c_str());
+    logger->debug("mlabns url: %s", url.c_str());
+    http_get(url,
+        [callback, logger](Error error, http::Response response) {
+            if (error) {
+                callback(error, Reply());
+                return;
+            }
+            if (response.status_code != 200) {
+                callback(UnexpectedHttpStatusCodeError(), Reply());
+                return;
+            }
+            Reply reply;
+            try {
+                auto node = json::parse(response.body);
+                reply.city = node.at("city");
```

Comment:
I am *not* happy with this diff. You *should not* have run clang-format in this case. This complicated the diff and forced me to do comparison by hand, locally, between two branches. This led me to discover that, in addition to reformatting everything, you also applied the following changes:

```diff
-                          reply.city = node["city"];
-                          reply.url = node["url"];
-                          for (auto ip2 : node["ip"]) {
+                reply.city = node.at("city");
+                reply.url = node.at("url");
+                for (auto ip2 : node.at("ip")) {
```

Why did you did that? Why did you hide that into a reformatted diff? Why did you not document that? I understand that you probably did this for what is probably a good reason, but I don't think that's the correct way of proceeding.

By producing diff complex to easily review without a good reason, you add burden on who is reviewing the diff, and I have the feeling that proceeding like this is the way to add unnoticed bugs to an implementation.

Could you please revert back everything to the old behavior such that the diff it's just the *obvious* moving of code from a source file to an header without any other change?